integrations/shopify/src/rest.ts

```diff
@@ -0,0 +1,311 @@
+  }
+
+  /**
+   * Fetch the number of resources of a given type.
+   */
+  async count(
+    key: string,
+    params?: Optional<OmitIndexSignature<Parameters<TResource["count"]>[0]>, "session">
+  ): CountReturnType {
+    return this.runTask(
+      key,
+      async (client, task, io) => {
+        const countResponse = await client.rest[this.resourceType].count(
+          this.#withSession(params ?? {})
+        );
+
+        const CountResponseSchema = z.object({
+          count: z.number(),
+        });
+
+        const parsed = CountResponseSchema.safeParse(countResponse);
+
+        if (!parsed.success) {
+          return JSON.parse(JSON.stringify(countResponse));
+        }
+
+        task.outputProperties = [
+          {
+            label: "Total",
+            text: String(parsed.data.count),
+          },
+        ];
+
+        return parsed.data;
+      },
+      {
+        name: `Count ${this.resourceType}s`,
+        params,
+      }
+    );
+  }
+
+  /**
+   * Create or update a resource of a given type. The resource will be created if no ID is specified.
+   */
+  async save<TFromData extends ShopifyInputType[TResourceType], TUpdate extends boolean = true>(
+    key: string,
+    params: {
+      update?: TUpdate;
+      fromData: TFromData;
+      session?: Session;
+    }
+  ): SaveReturnType<TResource, TUpdate, TFromData> {
+    return this.runTask(
+      key,
+      async (client, task, io) => {
+        const resource = new client.rest[this.resourceType](this.#withSession(params));
+
+        // mutate resource object with upserted data by default
+        await resource.save({ update: params.update ?? true });
+
+        return JSON.parse(JSON.stringify(resource));
+      },
+      {
+        name: `Upsert ${this.resourceType}`,
+        params,
+        properties: [
+          ...(params.fromData.id ? basicProperties({ id: params.fromData.id }) : []),
+          {
+            label: "Action",
+            text: params.fromData.id ? "Update" : "Create",
+          },
+        ],
+      }
+    );
+  }
+
+  /**
+   * Delete an existing resource.
+   */
+  async delete(
+    key: string,
+    params: Optional<Parameters<TResource["delete"]>[0], "session">
+  ): DeleteReturnType {
+    return this.runTask(
+      key,
+      async (client, task, io) => {
+        await client.rest[this.resourceType].delete(this.#withSession(params));
+        return;
+      },
+      {
+        name: `Delete ${this.resourceType}`,
+        params,
+        properties: basicProperties(params),
+      }
+    );
+  }
+}
+
+export class Rest {
+  constructor(
+    private runTask: ShopifyRunTask,
+    private session: Session
+  ) {}
+}
+
+interface MergeProxyConstructor {
+  new <TTarget extends Record<any, any>, TResult extends Record<any, any>>(
+    target: TTarget,
+    handler: ProxyHandler<TResult>
+  ): TTarget & TResult;
+}
+
+type ResourceMap = {
+  [KResourceType in ResourcesWithStandardMethods]: Resource<KResourceType>;
```

Comment:
Only a subset of Shopify Resources expose all standard methods. Need the raw client for ones that don't.